File: ChatGPT/python/output/output_24a.py
# ChatGPT-generated solution will go here.
import re
from collections import defaultdict, deque

def main():
    with open("input.txt") as f:
        lines = [line.strip() for line in f if line.strip()]

    wire_values = {}
    gates = []

    # Parse input
    for line in lines:
        if ':' in line:
            wire, val = line.split(':')
            wire_values[wire.strip()] = int(val.strip())
        else:
            m = re.match(r"(\w+) (AND|OR|XOR) (\w+) -> (\w+)", line)
            if m:
                in1, op, in2, out = m.groups()
                gates.append((in1, op, in2, out))

    # Continue processing gates until no changes occur
    pending = gates[:]
    while pending:
        next_round = []
        for in1, op, in2, out in pending:
            if in1 in wire_values and in2 in wire_values:
                a, b = wire_values[in1], wire_values[in2]
                if op == "AND":
                    wire_values[out] = a & b
                elif op == "OR":
                    wire_values[out] = a | b
                elif op == "XOR":
                    wire_values[out] = a ^ b
            else:
                next_round.append((in1, op, in2, out))
        if len(next_round) == len(pending):
            break  # no progress, avoid infinite loop
        pending = next_round

    # Gather all wires starting with 'z' and compute output value
    z_wires = sorted([k for k in wire_values if k.startswith('z')])
    binary_str = ''.join(str(wire_values[z]) for z in z_wires[::-1])
    print(int(binary_str, 2))

if __name__ == "__main__":
    main()

File: ChatGPT/clojure/output/output_24a.clj
;; ChatGPT-generated solution will go here.
(ns output_24a
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(defn parse-input [lines]
  (let [[inits gates] (split-with #(str/includes? % ":") lines)
        initial (into {}
                      (map (fn [line]
                             (let [[k v] (str/split line #": ")]
                               [k (parse-long v)]))
                           inits))
        gate-list (map (fn [line]
                         (let [[lhs out] (str/split line #" -> ")
                               [a op b] (str/split lhs #" ")]
                           {:op op :a a :b b :out out}))
                       gates)]
    [initial gate-list]))

(defn ready? [state a b]
  (and (contains? state a) (contains? state b)))

(defn compute [op va vb]
  (case op
    "AND" (if (and (= va 1) (= vb 1)) 1 0)
    "OR" (if (or (= va 1) (= vb 1)) 1 0)
    "XOR" (if (not= va vb) 1 0)))

(defn evaluate [initial gates]
  (loop [state initial
         remaining gates]
    (let [[ready not-ready] (split-with #(ready? state (:a %) (:b %)) remaining)
          new-state (reduce (fn [s {:keys [a b op out]}]
                              (assoc s out (compute op (s a) (s b))))
                            state
                            ready)]
      (if (empty? not-ready)
        new-state
        (recur new-state not-ready)))))

(defn part [lines]
  (let [[initial gates] (parse-input lines)
        state (evaluate initial gates)
        z-wires (->> state
                     (filter (fn [[k _]] (str/starts-with? k "z")))
                     (sort-by (fn [[k _]]
                                (parse-long (subs k 2)))))
        bits (mapv second z-wires)]
    (reduce (fn [acc b] (+ (* acc 2) b)) 0 bits)))

(defn -main [& args]
  (let [filename (first args)
        lines (str/split-lines (slurp filename))]
    (println (part lines))))
